### ISSUE TITLE
meteor 1.7  _includePaths is not iterable when importing scss file from node_modules

### ISSUE THREAD

**Issue by macrozone** [fourseven/meteor-scss#273]
- [ ] Feature request
- [x] Bug report
- [ ] Question

`meteor` version: 1.7-rc.11

`fourseven:scss` version: 4.9.0

I added a symlink to my node_module in /imports and when i start meteor, i get the error

` _includePaths is not iterable`

**Comment by copleykj** [comment]
So what you have to do is to symlink to the package folder in your node_modules folder, and then change your scss import to point to the asset within that folder..

for example this is how I've linked to the scss file within bootstrap_sass
```sh
$ cd imports
$ ln -s ../node_modules/bootstrap-sass/ .
```
Then in from my main.scss inside my client folder...

```scss
@import '../imports/styles/bootstrap-sass/assets/stylesheets/_bootstrap.scss';
```

**Comment by Floriferous** [comment]
I'm on 1.7.1-beta.19 and this still seems to happen, shouldn't we be able to import CSS directly from `node_modules`?

**Comment by narigondelsiglo** [comment]
1.7.0.5 still happens

**Comment by domq** [comment]
If you come here by Googling for "_includePaths is not iterable": [This appears to be a reload-related issue](https://forums.meteor.com/t/sass-in-local-package-from-node-modules/45865/4). Stopping and restarting Meteor helped in our hacking session today — Give it a try.

**Comment by alirezamirsepassi** [comment]
Same issue here. Meteor v1.8.0.2

**Comment by paul8046** [comment]
I can confirm that this is a reload-related issue. Stopping and restaring Meteor solves the problem.

**Comment by nikhilgoswami** [comment]
Correct answer as of Meteor 1.8 is 
in terminal
cd import folder in project root
ln -s ../node_modules/bootstrap-scss \\note path to package not individual file;
then in @import '../import/bootstrap-scss/assets/stylesheets/_bootstrap.scss';
Hope that will be it

**Comment by derwaldgeist** [comment]
@nikhilgoswami No, this doesn't solve it. It seems to be related to multiple SCSS files referencing another SCSS file with the same name, e.g. `variables.scss`, even if both are different.

**Comment by nikhilgoswami** [comment]
> @nikhilgoswami No, this doesn't solve it. It seems to be related to multiple SCSS files referencing another SCSS file with the same name, e.g. `variables.scss`, even if both are different.

Check if you have created sim link properly ln -s ' ../node_modules' if properly created you should see it in visual studio code even doing it multiple times creates problems

**Comment by derwaldgeist** [comment]
Everything is setup correctly.

The problem seems to be in either Meteor's or `fourseven:scss`'s compiler.

It happens if two SCSS files contain the same `@import` reference without a folder prefix, e.g.

`@import 'variables';`

In this case, Meteor seems to always take the file from the very first occurrence, which breaks the subsequent packages. If you do this:

`@import './variables';`

the problem is gone. However, there's a lot of 3rd party packages (like bootstrap) out there that do not import like this.

**Comment by nikhilgoswami** [comment]
> Everything is setup correctly.
> 
> The problem seems to be in either Meteor's or `fourseven:scss`'s compiler.
> 
> It happens if two SCSS files contain the same `@import` reference without a folder prefix, e.g.
> 
> `@import 'variables';`
> 
> In this case, Meteor seems to always take the file from the very first occurrence, which breaks the subsequent packages, or it breaks with the error above. If all files do import like this:
> 
> `@import './variables';`
> 
> the problem is gone. However, there's a lot of 3rd party packages out there (like bootstrap or fontawesome) that do not import like this.

Did you sim link individual file from package? Remember in my solution I told the to sim link entire package not individual file like ln -s '../node_modules/bootstrap-sass'

**Comment by derwaldgeist** [comment]
Yes, I symlinked the whole package.

**Comment by jamauro** [comment]
@derwaldgeist were you able to resolve? I just ran into this issue as well.

**Comment by derwaldgeist** [comment]
IIRC, I ended up patching the problematic import statements in these packages. Which is pretty annoying.

**Comment by nikhilgoswami** [comment]
You are right @derwaldgeist  I too am having hard time to include sass file I now converted all sass files to less in my project and will be using less for next projects.

**Comment by tr4g** [comment]
3 years and still no fix? wow I guess

**Comment by copleykj** [comment]
If anyone can provide a minimal reproduction so this issue can be properly triaged, I'm happy to reopen.

**Comment by derwaldgeist** [comment]
I ended up symlinking the folders from node_modules to import.

**Comment by tr4g** [comment]
@derwaldgeist it didn't work for me and it's utterly stupid a solution

**Comment by derwaldgeist** [comment]
It's not a nice solution, but at least it worked (on macOS; not sure if it works on Windows).

BTW: `fourseven:scss` has an even bigger issue related to imports: If the 3rd party library uses includes like these:

```@import "variables";```

instead of

```@import "./variables";```

it might cause conflicts with other files. It seems if `fourseven:scss` uses a cache for these files, and if two includes use the same import directive, they override each other. In these cases, I even had to copy the library and fix the imports manually.

**Comment by tr4g** [comment]
yeah it's time to switch to firebase, definitely

**Comment by copleykj** [comment]
@tr4g this is a third party, open source piece of software. Someone created it and generously put it into the world and allowed you to use it ,FOR FREE, in the hopes would save you time and effort. It is currently being maintained by members of the community who generously devote their time to help keep it, and numerous other packages, updated and available to you, FOR FREE. If you would like to join our efforts, by providing a viable minimal reproduciton, or even better a PR that fixes the issue, I'm more than happy to move things in the right direction.  If however all you have to contribute are complaints and criticism, then I honestly wish you the best in your journey to firebase.

**Comment by sebakerckhof** [comment]
@tr4g yes, please go and use firebase

**Comment by tr4g** [comment]
@copleykj you opensource kids always complain about people criticizing your products, like free has to equal crap, and that you only get quality when you pay. 

3 years to fix a problem that needs a hacky workaround like symlinks, to do something expected to just work, is nonsense

but sure, it's free

**Comment by sebakerckhof** [comment]
@tr4g Have you ever worked professionally as a software developer? Doesn't sound like it, since you'd understand the compromise that is software development. Every software has bugs that affect only a subset of users and are therefore deemed too low priority to fix in the available time. In the meantime this works fine for the 99% of folks not experiencing this bug. That's true for any software at every company.

So yeah, free doesn't need to equal crap, but free surely means we don't need to listen to your crap.

**Comment by sebastian-ruiz** [comment]
I also had the problem of _includePaths is not iterable

What I did is as follows to fix it:

1. got to `imports/ui/packages` and symlink all the node_modules that have scss:

```
ln -s ../../../node_modules/ag-grid-community
ln -s ../../../node_modules/bootstrap        
ln -s ../../../node_modules/@owczar/dashboard-style--airframe 
ln -s ../../../node_modules/font-awesome
ln -s ../../../node_modules/rc-slider   
ln -s ../../../node_modules/react-big-calendar  
ln -s ../../../node_modules/react-bootstrap-table-next 
ln -s ../../../node_modules/react-bootstrap-typeahead 
ln -s ../../../node_modules/react-datepicker         
ln -s ../../../node_modules/react-grid-layout 
ln -s ../../../node_modules/react-image-crop 
ln -s ../../../node_modules/react-quill     
ln -s ../../../node_modules/react-toastify 
ln -s ../../../node_modules/react-toggle
```

2. For me react-toastify was especially broken and I had to do the following.
In `imports/ui/packages/react-toastify/scss/main.scss` make the following change:

```
@import "{}/imports/ui/packages/react-toastify/scss/_variables.scss"; 
// @import "variables";
```

3. In the root of your app:
```
mkdir packages
cd packages
git clone https://github.com/sebastian-ruiz/meteor-scss
```
This is a fork of meteor-scss I wrote that fixes this issue.

I haven't worked with meteor in a year and I didn't do a pull-request for this change yet. I will do some more checks and then do a pull-request.

**Comment by apendua** [comment]
For what it's worth, this thing only happened to me when I included `node_modules` in `.meteorignore` file.

**Comment by derwaldgeist** [comment]
@sebastian-ruiz Such a PR would be awesome!

Regarding the `variables` import: I guess this is caused by another problem `fourseven:scss` has: If a path contains a filename only, you get conflicts if that filename appears in another package. It won't happen if you import it like this:

```
@import "./variables";
```

 I had to do this for quite a number of libraries. That's another reason why I tend to duplicate the files into my project instead of importing them directly from `node_modules`. Of course, this makes it much harder to do updates, because you always have to think about the patches you have to apply again.
